File: webapp/libfgg/pool/pool.ts
import FileHash from "./hash"
import { DataChunk, Meta, Hash } from "./data"
import { IFile } from "./file/file"

export default class Pool {
  // htmlDOMfile
  //sender: File | null = null
  sender: IFile | null = null

  // https://developer.mozilla.org/en-US/docs/Web/API/WritableStream/getWriter
  //recver: File | FileDigester | WritableStreamDefaultWriter | null = null
  recver: IFile | null = null

  fileHash: FileHash = new FileHash()

  meta: Meta | null = null
  hash: Hash | null = null

  doneCount: number = 0
  nextCount: number = 0

  OnFinish: () => void = () => {}
  OnProgress: (c :number) => void = (_) => {}

  // [safari] max-message-size: 64 * 1024
  // [chrome, firefox] max-message-size: 256 * 1024
  chunkSize: number = 32 * 1024

  currentSize: number = 0
  pendingSize: number = 0

  setSend(file: IFile) {
    this.sender = file
  }

  setRecv(file: IFile) {
    this.recver = file
  }

  async recvMeta(meta: Meta): Promise<void> {
    this.meta = meta
    await this.recver?.setMeta(meta)
  }

  async sendMeta(): Promise<Meta> {
    if (!this.sender) {
      throw "Not found sender file"
    }

    const meta = await this.sender.getMeta()
    this.meta = meta
    return meta
  }

  sendHash(): Hash {
    if (!this.sender) {
      throw "Not found sender file"
    }

    return {
      name: this.meta?.name || "",
      hash: this.fileHash.sum(),
    }
  }

  recvHash(hash: Hash): boolean {
    return hash.hash === this.fileHash.sum()
  }

  async sendData(c: DataChunk): Promise<ArrayBuffer> {
    if (!this.sender) {
      throw "Not found sender file"
    }

    const data = await this.sender.read(c.offset, c.length)

    this.fileHash.onData(c, data)
    this.OnProgress(this.fileHash.offset)
    return data
  }

  async recvData(c: DataChunk, data: ArrayBuffer): Promise<void> {
    if (!this.recver) {
      throw "Not found recver file"
    }

    this.currentSize += c.length
    this.fileHash.onData(c, data)
    this.OnProgress(this.fileHash.offset)

    return this.recver.write(c.offset, c.length, data)
  }

  next(): DataChunk | null {
    if (!this.meta) {
      throw "Not found recver file"
    }

    if (this.currentSize >= this.meta.size) {
      this.OnFinish()
      return null
    }

    if (this.pendingSize >= this.meta.size) {
      return null
    }

    let length = this.chunkSize
    const next = this.currentSize + this.chunkSize
    if (next > this.meta.size) {
      const n = next - this.meta.size
      length = this.chunkSize - n
    }

    const offset = this.pendingSize

    this.pendingSize += this.chunkSize
    return {
      offset: offset,
      length: length,
    }
  }
}

File: libfgg/pool/pool.go
package pool

import (
	"errors"
	"fmt"
	"os"
	"sync"
)

type Meta struct {
	Name string `json:"name"`
	Type string `json:"type"`
	Size int64  `json:"size"`
}

type Hash struct {
	File string `json:"name"`
	Hash string `json:"hash"`
}

type Pool struct {
	sender *os.File
	recver *os.File

	fileHash *fileHash

	meta *Meta
	hash *Hash

	doneCount int
	nextCount int

	mu sync.Mutex

	OnFinish   func()
	OnProgress func(c int64)

	chunkSize int64

	currentSize int64
	pendingSize int64
}

func New() *Pool {
	return &Pool{
		fileHash:   newFileHash(),
		OnFinish:   func() {},
		OnProgress: func(c int64) {},
		chunkSize:  32 * 1024,
	}
}

func (p *Pool) SetSend(file string) (err error) {
	p.sender, err = os.Open(file)
	return
}

func (p *Pool) SetRecv(file string) (err error) {
	p.recver, err = os.Create(file)
	return
}

func (p *Pool) SendMeta() (*Meta, error) {
	if p.meta != nil {
		return p.meta, nil
	}

	if file := p.sender; file != nil {
		// TODO: Maybe use https://github.com/gabriel-vasile/mimetype
		mimeType, _ := getFileContentType(file)
		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
		// mimeType := "application/octet-stream"

		stat, err := file.Stat()
		meta := &Meta{
			Name: file.Name(),
			Type: mimeType,
			Size: stat.Size(),
		}
		p.meta = meta
		return meta, err
	}

	return nil, errors.New("Sender not found file")
}

func (p *Pool) RecvMeta(meta *Meta) error {
	var err error
	if p.recver == nil {
		p.recver, err = os.Create(meta.Name)
	}
	p.meta = meta
	return err
}

func (p *Pool) SendHash() (*Hash, error) {
	return &Hash{
			File: p.sender.Name(),
			Hash: p.fileHash.sum(),
		}, func() error {
			if p.fileHash.offset != p.meta.Size {
				return errors.New("Not sum")
			}
			return nil
		}()
}

func (p *Pool) RecvHash(meta *Hash) error {
	if meta.Hash == p.fileHash.sum() {
		return nil
	}
	return fmt.Errorf("'%s' Not match '%s'", meta.Hash, p.fileHash.sum())
}
